### PR TITLE
feat: ✨ sd-video

### PR DESCRIPTION
**Pull request by abudd1094**
adds sd-video component: https://github.com/solid-design-system/solid/issues/281

Please see my comment on the ticket above for status information

### CODE REVIEW

File: packages/components/src/components/video/videoExample.ts
Comment (mariohamann) [inline packages/components/src/components/video/videoExample.ts:54]: This should be part of a sample. This whole file shouldn't exist, as this is not a component.

In the sample We should wire up plain JS with the component.

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:55]: The play icon should become a system icon and set as default slot content here.

File: packages/components/src/components/video/videoExample.ts
Comment (mariohamann) [inline packages/components/src/components/video/videoExample.ts:53]: Please remove.

**Review by mariohamann — CHANGES_REQUESTED** [review]

File: packages/components/src/components/video/videoExample.ts
Comment (abudd1094) [reply packages/components/src/components/video/videoExample.ts:54]: is there any other example in the repo to look to?

File: packages/components/src/components/video/video.ts
Comment (abudd1094) [reply packages/components/src/components/video/video.ts:55]: What does "become a system icon" mean?

Comment (abudd1094) [reply packages/components/src/components/video/video.ts:55]: I assume you want:
`<sd-icon library="global-resources" name="system/start" color="primary" class="text-[4rem]"></sd-icon>`

To become something like:
`<sd-icon library="system" name="start" color="primary" class="text-[4rem]"></sd-icon>`

But that doesn't work.  Please provide further instructions

File: packages/components/src/components/video/videoExample.ts
Comment (abudd1094) [reply packages/components/src/components/video/videoExample.ts:53]: remove what exactly?  This whole component will be removed as per the comment above

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [reply packages/components/src/components/video/video.ts:55]: As soon as you merge main, it should work now as @Vahid1919 just merged an update of https://github.com/solid-design-system/solid/blob/main/packages/components/src/components/icon/library.system.ts in his latest commit to main.

File: packages/components/src/components/video/video.stories.ts
Comment (karlbaumhauer) [inline packages/components/src/components/video/video.stories.ts:14]: should we add some information here about the non-existing sd-media component? Or maybe leave it out of here and add it as soon as the component is ready.

File: packages/components/src/components/video/video.ts
Comment (karlbaumhauer) [inline packages/components/src/components/video/video.ts:10]: pls change before merging

**Review by karlbaumhauer — CHANGES_REQUESTED** [review]

File: packages/components/src/components/video/video.stories.ts
Comment (mariohamann) [reply packages/components/src/components/video/video.stories.ts:14]: +1 – would remove everything related `sd-media`

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [reply packages/components/src/components/video/video.ts:10]: You mean the `status`?

**Comment by mariohamann** [comment]
@solid-design-system/design 

I know, that the play button is mathematically correct centered but it feels visually off.

CleanShot 2023-10-11 at 09 18 34@2x

I would propose to add a 4px margin to the left to make it visually more pleasing:

CleanShot 2023-10-11 at 09 20 11@2x

Apparently in Figma there already is a little margin to the left in the icon itself:

CleanShot 2023-10-11 at 09 22 55@2x

@Vahid1919 Do you remember, where the icon was coming from?

**Comment by mariohamann** [comment]
The default story should contain an example where you can play around with the knobs. We should add a placeholder image in format 16:9 for that in the background. 

I just opened a PR for the image: https://github.com/solid-design-system/solid/pull/457

**Comment by van-nguyen-ht** [comment]
@mariohamann yes indeed, in [figma](https://www.figma.com/file/VTztxQ5pWG7ARg8hCX6PfR/Solid-DS-%E2%80%93-Component-Library?type=design&node-id=1663-18047&mode=design&t=YPSuBz0TT42I9sPK-11) the icon has a padding to the left

File: packages/components/src/components/video/video.stories.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.stories.ts:37]: As mentioned here https://github.com/solid-design-system/solid/pull/451#issuecomment-1757026290 we should show at least an image in the main slot.

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:36]: If we provide this, we should have a second method, that does the opposite.

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:18]: ❤️

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:27]:
```suggestion
  /** Set to `true` to hide the play icon and the overlay. */
```

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:30]:
```suggestion
  /** Set to `true` to show a dark overlay. Only used when `playing` is `false`. */
```

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:57]: "transition" === 😍

File: packages/components/src/components/video/video.stories.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.stories.ts:38]: When implementing the placeholder, make another story which does `Playing x Overlay`.

**Review by mariohamann — CHANGES_REQUESTED** [review]

File: packages/components/src/components/video/video.ts
Comment (abudd1094) [reply packages/components/src/components/video/video.ts:10]: what should status be?

Comment (karlbaumhauer) [reply packages/components/src/components/video/video.ts:10]: status and version.

Status should be stable and version needs to be updated to the latest +1 right before merging.

File: packages/components/src/components/video/video.ts
Comment (abudd1094) [reply packages/components/src/components/video/video.ts:36]: I don't think it makes sense to expose this or add a "pause" method, control is always coming from the wrapped component and should be reflected via the property.  

I propose marking this function as private and using only for the click event to make it simple.  External control should use the "playing" property exclusively.

**Comment by abudd1094** [comment]
> @solid-design-system/design
> 
> I know, that the play button is mathematically correct centered but it feels visually off.
> 
> CleanShot 2023-10-11 at 09 18 34@2x
> I would propose to add a 4px margin to the left to make it visually more pleasing:
> 
> CleanShot 2023-10-11 at 09 20 11@2x
> Apparently in Figma there already is a little margin to the left in the icon itself:
> 
> CleanShot 2023-10-11 at 09 22 55@2x
> @Vahid1919 Do you remember, where the icon was coming from?

I had a translate on there I thought you said to "remove".  Margin will distort the background:
image

I will re-add the translate

File: packages/components/src/components/video/video.ts
Comment (karlbaumhauer) [reply packages/components/src/components/video/video.ts]: good point... @mariohamann what do you think?

File: packages/components/src/components/video/video.ts
Comment (karlbaumhauer) [inline packages/components/src/components/video/video.ts:11]:
```suggestion
 * @since 1.19.0
```

**Review by karlbaumhauer — CHANGES_REQUESTED** [review]

**Comment by Vahid1919** [comment]
> @Vahid1919 Do you remember, where the icon was coming from?

It was from the icon-library from the previous Union Investment project I believe.

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [reply packages/components/src/components/video/video.ts]: Ok. Either both or none. In accordion we provide it like this:

CleanShot 2023-10-11 at 13 23 20@2x

But I think it's totally valid to just work with the prop here.

**Comment by mariohamann** [comment]
> > @Vahid1919 Do you remember, where the icon was coming from?
> 
> It was from the icon-library from the previous Union Investment project I believe.

It seems like the icon is not correctly set in it's frame. Could you please add a ticket, that we have to include its padding and remove the added translate by Alec then?

**Comment by mariohamann** [comment]
I could get the video playing when doing it via keyboard.

https://64469bd84c5ea1ce2dd5d016-nziojohwnh.chromatic.com/?path=/story/components-sd-video--video-element

**Comment by mariohamann** [comment]
So we should have one default story with jsut the imag e + one Overlay x Playing Story. @abudd1094  :)

**Comment by Vahid1919** [comment]
I noticed that the default play button is still visible underneath our own for Firefox --> 
image

PS: This only seems to be the case for the first button press, it does not come back afterwards.

**Comment by abudd1094** [comment]
> I could get the video playing when doing it via keyboard.
> 
> https://64469bd84c5ea1ce2dd5d016-nziojohwnh.chromatic.com/?path=/story/components-sd-video--video-element
> 
> And I'm not sure regarding the focus styles – I think we should use our default ones.

By "default" you mean: focus-visible:focus-outline

**Comment by mariohamann** [comment]
@karlbaumhauer aybe we should reintroduce the `poster` slot – this would prevent everything here?

**Comment by karlbaumhauer** [comment]
> @karlbaumhauer aybe we should reintroduce the `poster` slot – this would prevent everything here as it could just be an overlay over the video?

thats a good idea...

**Comment by abudd1094** [comment]
@karlbaumhauer 

- I removed my translate for the icon as requested, Vahid's PR should fix that without further changes in the sd-video.
- The play icon is centered in our component, which does not take into account the "controls" in the underlying video component.  

I'll add a poster slot, we'll have a quick check in the daily, and I'll add stories / tests if that is accepted.

**Comment by mariohamann** [comment]
@abudd1094 @karlbaumhauer  An alternative could be to only show controls when playing. This could work when you extend your JS, wouldn't it?

**Comment by abudd1094** [comment]
@karlbaumhauer @mariohamann 

OK, after much pain and learning, sd-video is ready for another review.

I have added a "poster" slot which seeks to mirror the behavior of the "poster" attribute on the native "video" element.  It will be shown under the optional overlay and on top of the default video element slot with no pointer events.  So the user must click (or use keyboard to activate) the "play-icon" for the initial play, after which the "poster" will be hidden, and the user can click through the overlay to the video controls.  

I have added as many tests as I could think to add.  I would also like to add the basic "accessibility" test, but I'm not sure I understand how it works.  It seems to fail for Chrome here but I cannot figure out the error:
image

Accessibility has been further bolstered and I added a story for "mouseless" control.  Please advise.

**Comment by karlbaumhauer** [comment]
@abudd1094 sorry for the delay... I will have a look into it by tomorrow.

File: packages/components/src/components/video/video.stories.ts
Comment (karlbaumhauer) [inline packages/components/src/components/video/video.stories.ts:84]: I like that one...!

File: packages/components/src/components/video/video.stories.ts
Comment (karlbaumhauer) [inline packages/components/src/components/video/video.stories.ts:141]:
```suggestion
 * Here we wrap a `video` element that includes setting an image to the `poster` slot.
```

File: packages/components/src/components/video/video.stories.ts
Comment (karlbaumhauer) [inline packages/components/src/components/video/video.stories.ts:165]: we need to deactivate the mouse in this story, when hovering over the component. Have a look into the sd-tag mouseless story to see how its implemented.

**Review by karlbaumhauer — CHANGES_REQUESTED** [review]

**Comment by karlbaumhauer** [comment]
> @karlbaumhauer @mariohamann
> 
> OK, sd-video is finally ready for another review.
> 
> I have added a "poster" slot which seeks to mirror the behavior of the "poster" attribute on the native "video" element. It will be shown under the optional overlay and on top of the default video element slot with no pointer events. So the user must click (or use keyboard to activate) the "play-icon" for the initial play, after which the "poster" will be hidden, and the user can click through the overlay to the video controls.
> 
> I have added as many tests as I could think to add. I would also like to add the basic "accessibility" test, but I'm not sure I understand how it works. It seems to fail for Chrome here but I cannot figure out the error:
> 
> Accessibility has been further bolstered and I added a story for "mouseless" control. Please advise.

So the component looks good and I just found some small text changes and 1 issue in the mouseless story.

Regarding the tests, I unfortunately cannot help here. @Vahid1919 or @mariohamann pls advise if possible.

**Comment by mariohamann** [comment]
CleanShot 2023-10-18 at 10 18 00@2x

Could you please set the size on the image to prevent this?

I think you're good to go with `class="w-WHATEVER-WIDTH aspect-video cover"`.

**Comment by mariohamann** [comment]
Please use one of our internal images as poster. As these will be used for screenshot tests, we shouldn't have any external dependencies on them. (That's why we e. g. intercept all calls to global-resources and provide own icons, as we don't want to have any external stuff.) 

CleanShot 2023-10-18 at 10 20 47@2x

File: packages/components/src/components/video/video.stories.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.stories.ts:16]: You can use here Tailwind as well. Something like `w-ANYWIDTH aspect-video cover`.

File: packages/components/src/components/video/video.stories.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.stories.ts:112]: Could you add some more documentation for the users of our library?

Pointing them to what's important for them etc. Some stuff isn't important, like Storybook changes, maybe you can clarify that, too.

So more thinking from a user of our library: What is importatnt. What do I have to do why. What's not important.

File: packages/components/src/components/video/video.stories.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.stories.ts:153]:
```suggestion
      <div id="sd-video-example">
```

File: packages/components/src/components/video/video.test.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.test.ts:12]: Especially here: Definitely remove external dependencies. If needed just inline the image with base64 or whatever.

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:85]:
```suggestion
            'bg-black/65 w-full h-full absolute top-0 left-0 pointer-events-none z-20 play-pause-transition'
```

Couldn't you use something like bg-black/65 or something?

File: packages/components/src/components/video/video.ts
Comment (mariohamann) [inline packages/components/src/components/video/video.ts:96]:
```suggestion
            'absolute left-1/2 top-1/2 translate-x-1/2 translate-y-1/2 p-4 bg-white bg-opacity-75 rounded-full flex items-center justify-center z-30 play-pause-transition'
```

Something like this should work.

File: packages/components/src/components/video/video.stories.ts
Comment (abudd1094) [reply packages/components/src/components/video/video.stories.ts:112]: You mean via code comments in this block?

File: packages/components/src/components/video/video.ts
Comment (abudd1094) [reply packages/components/src/components/video/video.ts:96]: Before:
image

After:
image

File: packages/components/src/components/video/video.ts
Comment (abudd1094) [reply packages/components/src/components/video/video.ts:85]: discussed via phone --- no bg-black token exists :(

File: packages/components/src/components/video/video.stories.ts
Comment (mariohamann) [reply packages/components/src/components/video/video.stories.ts:112]: Yes

**Comment by mariohamann** [comment]
> > Please use one of our internal images from our placeholders package as poster. As these will be used for screenshot tests, we shouldn't have any external dependencies on them. (That's why we e. g. intercept all calls to global-resources and provide own icons, as we don't want to have any external stuff.)
> > CleanShot 2023-10-18 at 10 20 47@2x
> 
> What about for the video?

As this doesn't show up in the screenshots, I'm less opinionated there. We maybe should have a default there, too. But for the moment I think that's fine.

File: packages/components/src/components/video/video.test.ts
Comment (abudd1094) [reply packages/components/src/components/video/video.test.ts:12]: "just inline the image with base64 or whatever" --- please expand on what you mean.  I don't think you want:
image
I have removed the src completely, it still passes the tests.  The video remains in the tests and is an external dependency.  If you have a better way to do it, please explain, otherwise I'll leave it like this

**Comment by abudd1094** [comment]
All comments have been addressed.  @mariohamann I added this issue regarding the accessibility test:
https://github.com/solid-design-system/solid/issues/483

Please let me know if I should refine this ticket or make further changes.

File: packages/components/src/components/video/video.stories.ts
Comment (karlbaumhauer) [inline packages/components/src/components/video/video.stories.ts:27]: @mariohamann was this `w-WHATEVER-WIDTH` really what you expected? If yes, why? Would love to then see a description here.

**Review by karlbaumhauer — APPROVED** [review]

**Comment by karlbaumhauer** [comment]
> All comments have been addressed. @mariohamann I added this issue regarding the accessibility test: #483
> 
> Please let me know if I should refine this ticket or make further changes.

@abudd1094 I looked at the ticket and re-ordered it a bit. thx for creating it.

**Comment by abudd1094** [comment]
@van-nguyen-ht 
image

Point 1:
It is pixel for pixel the same, 96 x 96, please check the measurements by number:
image
image

Point 2:
"There's no interactions with the play button" --- yes, that is expected.  In the default story, the component uses a placeholder image as its "default slot" so there is no video to play.  Please see the Video Element stories for interaction demonstrations.

**Comment by coraliefeil** [comment]
SRY messed up assignees by accident

**Comment by abudd1094** [comment]
@van-nguyen-ht requested that I implement the queries from the design:
image

@mariohamann I didn't see 414px as a standard breakpoint, it is implemented in the css block.  If there is a more desirable way to implement this or you want to adjust the format please advise.

**Comment by EllaELSA** [comment]
Not sure where to add a comment, here or in chromatic? 
Shouldn't we provide a hover effect on video or at least at the play button?

**Comment by coraliefeil** [comment]
Removed myself ... Van’s taking care of design review

**Comment by coraliefeil** [comment]
@abudd1094 could we use the sd-interactive for the play icon and the close icon? states are the same.

**Comment by abudd1094** [comment]
@EllaELSA 
This appears to be a different element entirely <sd-play-button> which is not mentioned in the ticket.

@coraliefeil 
What is sd-interactive?  I don't see it in Storybook or Figma... I do see a ticket in Review, perhaps it is not yet merged then.  In that case, we can use it, but I'll have to wait until it is merged.  

@mariohamann @karlbaumhauer @yoezlem 
Seems like we missed a dependency.  Should I add <sd-play-button> and include it in the default slot as part of this PR?

Sounds like what we want, is to merge sd-interactive, use it to build the sd-play-button, and then use that as the default slot for sd-video???

That is fine for me but this information was not transparent.  Please advise before I continue.

**Comment by coraliefeil** [comment]
@abudd1094 https://solid-design-system.fe.union-investment.de/x.x.x/storybook/?path=/docs/styles-sd-interactive--docs

**Comment by mariohamann** [comment]
I'm not sure if the `sd-interactive` class already works inside components. Maybe it does, maybe not. I would suggest du create a follow-up ticket for a possible refactoring, but as long as the current styles are fine, we should proceed.

**Comment by mariohamann** [comment]
@abudd1094 You can check an implemetnation example in sd-carousel:

import InteractiveStyles from '../../styles/interactive/interactive.css?inline';

You can sk @Vahid1919 for further implementation details.

**Comment by mariohamann** [comment]
@abudd1094 You could either set `poster` and `overlay` to pointer-events-none, so that clicking will be passed to the element or you set the click handler on the whole element.

@EllaELSA is right, that clicking anywhere should trigger playing: https://www.w3schools.com/html/html5_video.asp https://www.w3schools.com/tags/att_video_poster.asp

Besides that: Hover styles aren't defined anywhere for cover elements (as this isn't defined in design) and doesn't work in current UI videos as well. I suggest this to be an improvement if design catched up with everything.

**Comment by mariohamann** [comment]
@abudd1094 I don't understand the purpose of the `sd-play-button` as well, especially as should never handle pause with a big filling icon on the video?

I would definitely say to skip this component here and we will validate, if there's a reason for the `sd-play-button`. If yes, we might refactor at a later point, but for the moment, I don't see anything we can benefit from it.

**Comment by abudd1094** [comment]
@mariohamann @karlbaumhauer @EllaELSA @coraliefeil @van-nguyen-ht 
Play button now contains "sd-interactive", covers the whole element to field all click events and applies hover style to the play icon.  Click events have been fine tuned for precision.  I added a debounce functionality in the story to prevent flickering of our play icon on "seeking" events, without this, the icon will show up briefly every time a user initiates seeking in the video player.

**Comment by abudd1094** [comment]
77 comments, I'm going for a record...

@van-nguyen-ht @mariohamann 
I moved the media query from screen to element as mentioned on Figma.  This should cover all of the feedback we have gathered here.  Will update PR again when Quality Gate is fixed.

File: packages/components/src/components/navigation-item/navigation-item.ts
Comment (mariohamann) [inline packages/components/src/components/navigation-item/navigation-item.ts:14]: Is this intended?

**Review by mariohamann — APPROVED** [review]

File: packages/components/src/components/navigation-item/navigation-item.ts
Comment (abudd1094) [reply packages/components/src/components/navigation-item/navigation-item.ts:14]: yes, the version was incorrect --- regarding status, it is not clear to me how you determine that, imo this is now stable and not experimental

Comment (karlbaumhauer) [reply packages/components/src/components/navigation-item/navigation-item.ts:14]: everything is stable unless we specifically announce something as experimental

**Comment by abudd1094** [comment]
@van-nguyen-ht @coraliefeil @EllaELSA 
I have added you as assignees for the UI Tests at your lesiure.

@Vahid1919 @yoezlem @DanielHargesheimer 
I have adde you as assignees simply to have a look at this new feature (as discussed this morning in the daily).  No need to review!

**Pull request merged**
